tests/MongoDB.Driver.Tests/PipelineDefinitionTests.cs

```diff
@@ -34,9 +34,33 @@ public void Constructor_should_verify_the_inputs_and_outputs_of_the_stages_and_t
                 new BsonDocumentPipelineStageDefinition<BsonDocument, Person>(new BsonDocument())
             };
 
-            Action act = () => new PipelineStagePipelineDefinition<Person, Person>(stages);
+            var exception = Record.Exception(() => new PipelineStagePipelineDefinition<Person, Person>(stages));
+
+            var expectedException =
+                $"The input type to stage[2] was expected to be {typeof(Pet)}, but was {typeof(BsonDocument)}.";
+            exception.Should().BeOfType<ArgumentException>();
+            exception.Message.Should().Contain(expectedException);
+            ((ArgumentException) exception).ParamName.Should().Be("stages");
+        }
+
+        [Fact]
+        public void Constructor_should_verify_the_inputs_and_outputs_of_the_stages_and_throw_when_final_stage_is_invalid()
+        {
+            var stages = new IPipelineStageDefinition[]
+            {
+                new BsonDocumentPipelineStageDefinition<Person, BsonDocument>(new BsonDocument()),
+                new BsonDocumentPipelineStageDefinition<BsonDocument, Pet>(new BsonDocument()),
+                new BsonDocumentPipelineStageDefinition<Pet, BsonDocument>(new BsonDocument())
+            };
+
+            var exception = Record.Exception(() => new PipelineStagePipelineDefinition<Person, Person>(stages));
+
+            var expectedException =
+                $"The output type to the last stage was expected to be {typeof(Person)}, but was {typeof(BsonDocument)}.";
+            exception.Should().BeOfType<ArgumentException>();
+            exception.Message.Should().Contain(expectedException);
+            ((ArgumentException) exception).ParamName.Should().Be("stages");
```

Comment:
nit: normally these assertions would be written as:

    var e = exception.Should().BeOfType<ArgumentException>().Subject;
    e.ParamName.Should().Be("stages");
    e.Message.Should().Contain($"The output type to the last stage was expected to be {typeof(Person)}, but was {typeof(BsonDocument)}.");

Reply:
Fixed